conseil-lorre/src/main/scala/tech/cryptonomic/conseil/indexer/tezos/TezosGovernanceOperations.scala

```diff
@@ -0,0 +1,279 @@
+      proposalsBlocks: Map[Block, ProtocolId],
+      proposalsBallots: Map[Block, List[Voting.Ballot]],
+      levelsRolls: Map[Int, List[Voting.BakerRolls]]
+  )(
+      implicit ec: ExecutionContext
+  ): Future[List[GovernanceAggregate]] = {
+
+    //local functions, should simplify reading the calling code
+    def countBallotsPerLevel(block: Block) =
+      TezosDatabaseOperations.getBallotOperationsForLevel(block.data.header.level)
+
+    def countBallotsPerCycle(cycle: Int) =
+      TezosDatabaseOperations.getBallotOperationsForCycle(cycle)
+
+    // as stated, this is a runtime failure if the block contains the wrong metadata
+    def proposalOperationsHashes(block: Block) = block.data.metadata match {
+      case BlockHeaderMetadata(_, _, _, _, _, level) =>
+        TezosDatabaseOperations.getProposalOperationHashesByCycle(level.cycle)
+    }
+
+    logger.info("Searching for governance data in voting period...")
+
+    //main algorithm
+    val cycles = proposalsBlocks.keys
+      .filter(votingPeriodIn(ballotPeriods))
+      .map(_.data.metadata)
+      .collect {
+        case md: BlockHeaderMetadata => md.level.cycle
+      }
+      .toList
+      .distinct
+
+    val cycleBallotCountsResult =
+      Future
+        .traverse(cycles) { cycle =>
+          db.run(countBallotsPerCycle(cycle)).map(cycle -> BallotsPerCycle(_))
+        }
+        .map(_.toMap)
+
+    val levelBallotCountsResult =
+      Future
+        .traverse(
+          proposalsBlocks.keys.filter(votingPeriodIn(ballotPeriods))
+        ) { block =>
+          db.run(countBallotsPerLevel(block)).map(block -> BallotsPerLevel(_))
+        }
+        .map(_.toMap)
+
+    val proposalCountsResult = Future.traverse(
+      proposalsBlocks.keys.filter(votingPeriodIs(VotingPeriod.proposal))
+    ) { block =>
+      db.run(proposalOperationsHashes(block).map(block -> _))
+    }
+
+    for {
+      levelCountsMap <- levelBallotCountsResult
+      cycleCountsMap <- cycleBallotCountsResult
+      proposalCounts <- proposalCountsResult
+    } yield
+      fillAggregates(
+        proposalsBlocks,
+        levelsRolls,
+        proposalsBallots,
+        levelCountsMap,
+        cycleCountsMap,
+        proposalCounts.toMap
+      )
+  }
+
+  /* Having all data ready, we can process per block,
+   * extract the numbers returning a collector object.
+   */
+  private def fillAggregates(
+      proposalsBlocks: Map[Block, ProtocolId],
+      listingsPerLevel: Map[Int, List[Voting.BakerRolls]],
+      ballots: Map[Block, List[Voting.Ballot]],
+      ballotCountsPerLevel: Map[Block, BallotsPerLevel],
+      ballotCountsPerCycle: Map[Int, BallotsPerCycle],
+      proposalProtocolCounts: Map[Block, Map[ProtocolId, Int]] //comes from individual operations on the block
+  ): List[GovernanceAggregate] =
+    proposalsBlocks.toList.flatMap {
+      case (block, proposal) =>
+        val listing = listingsPerLevel.get(block.data.header.level).getOrElse(List.empty)
```

Comment:
you can change this to:
`listingsPerLevel.getOrElse(block.data.header.level, List.empty)`

Reply:
So right!